build-llvm.sh

```diff
@@ -68,21 +68,26 @@ if [ -z "$CHECKOUT_ONLY" ]; then
 fi
 
 if [ ! -d llvm-project ]; then
-    # When cloning master and checking out a pinned old hash, we can't use --depth=1.
-    git clone https://github.com/llvm/llvm-project.git
+    mkdir llvm-project
+    pushd llvm-project
+    git init
+    git remote add origin https://github.com/llvm/llvm-project.git
+    git fetch --depth 1 origin "$LLVM_VERSION"
+    git checkout FETCH_HEAD
+    popd
     CHECKOUT=1
 fi
 
 if [ -n "$SYNC" ] || [ -n "$CHECKOUT" ]; then
-    cd llvm-project
-    [ -z "$SYNC" ] || git fetch
-    git checkout $LLVM_VERSION
+    pushd llvm-project
+    [ -z "$SYNC" ] || git fetch --depth 1 origin "$LLVM_VERSION"
+    git checkout FETCH_HEAD
```

Comment:
Couldn't we keep `git checkout $LLVM_VERSION` here, or is there a reason why we can't? In case `$SYNC` wasn't set, then we do a `git checkout FETCH_HEAD` without executing the `git fetch` on the line right above. In practice this will happen if the if statement above was executed, so it'd work out fine, but I would prefer to only refer to `FETCH_HEAD` if it refers to things fetched in the line exactly preceding it.

With that in mind, could the `git checkout FETCH_HEAD` on line 76 be omitted too?

Reply:
About first your section - I think you're right. Git checkout FETCH_HEAD doesn't have a sense if we didn't git fetch. I don't know how to fix it to get it right. It's only good option if somebody wants to download llvmorg-12.0.0 and don't care about git log or updating (it could be problematic when you change llvmorg-12.0.0 to another version without sync).

> With that in mind, could the git checkout FETCH_HEAD on line 76 be omitted too?

Nope, the repository will be empty:
```console
$ git fetch --depth 1 origin "llvmorg-12.0.0"
 * tag                   llvmorg-12.0.0 -> FETCH_HEAD
$ git status
On branch master

No commits yet

nothing to commit (create/copy files and use "git add" to track)
$ ls | wc -l
0
```

Reply:
No, but after the clone/fetch in the `if [ ! -d llvm-project ]; then` block we set `CHECKOUT=1`, which triggers running the `if [ -n "$SYNC" ] || [ -n "$CHECKOUT" ]; then` block, which then checks out the desired version (and applies patches on top). So the first block would only clone/fetch things, and the second block does the checkout+patching.

I tried now though, and after a `git fetch --depth 1 origin "llvmorg-12.0.0"`, the `git checkout $LLVM_VERSION` doesn't work indeed, `FETCH_HEAD` has to be used. It does work if `$LLVM_VERSION` is a commit hash, but not for a tag name.

So that makes a bit of an issue, because I don't want to use `FETCH_HEAD` unless strictly the previous line did fetch that.

Is there any way to make the `git fetch --depth 1 origin "$LLVM_VERSION"` command actually sync the tag itself, but not all other tags? (Adding `--tags` makes it fetch all tags.)

If not, I think I'd prefer the code to look like this:

```
if [ ! -d llvm-project ]; then
    mkdir llvm-project 
    cd llvm-project
    git init 
    git remote add origin https://github.com/llvm/llvm-project.git
    git fetch --depth 1 origin "$LLVM_VERSION" 
    git checkout FETCH_HEAD
    cd ..
    CHECKOUT=1
fi

if [ -n "$SYNC" ] || [ -n "$CHECKOUT" ]; then
    cd llvm-project
    if [ -n "$SYNC" ]; then 
        git fetch --depth 1 origin "$LLVM_VERSION"
        git checkout FETCH_HEAD
    fi
```

That way the `FETCH_HEAD` always refers to a matching `git fetch` right above it. In this case the clone block also does the checkout, but that's probably tolerable.

Reply:
You can't use `git checkout $LLVM_VERSION`. Git fetch with depth=1 creates a short-lived ref.

There is another problem in your `prefer the code `. It will download every time all files from git (142 MB with compression). Here is example when I would like to change from 12.0.0 to 12.0.1-rc1
```console
root@ryz2:~# mkdir asdf
root@ryz2:~# cd asdf
root@ryz2:~/asdf#  git init 
Initialized empty Git repository in /root/asdf/.git/
root@ryz2:~/asdf# git remote add origin https://github.com/llvm/llvm-project.git
root@ryz2:~/asdf# git fetch --depth 1 origin "llvmorg-12.0.0"
remote: Enumerating objects: 105441, done.
remote: Counting objects: 100% (105441/105441), done.
remote: Compressing objects: 100% (90462/90462), done.
remote: Total 105441 (delta 18570), reused 59695 (delta 11503), pack-reused 0
Receiving objects: 100% (105441/105441), 142.42 MiB | 10.58 MiB/s, done.
Resolving deltas: 100% (18570/18570), done.
From https://github.com/llvm/llvm-project
 * tag                   llvmorg-12.0.0 -> FETCH_HEAD
root@ryz2:~/asdf# git checkout FETCH_HEAD
Updating files: 100% (100407/100407), done.
Note: switching to 'FETCH_HEAD'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at d28af7c65 remove -fpch-codegen and -fpch-debuginfo from Clang 12.0 release notes
root@ryz2:~/asdf# git fetch --depth 1 origin "llvmorg-12.0.1-rc1"
remote: Enumerating objects: 105454, done.
remote: Counting objects: 100% (105454/105454), done.
remote: Compressing objects: 100% (90486/90486), done.
remote: Total 105454 (delta 18567), reused 59698 (delta 11492), pack-reused 0
Receiving objects: 100% (105454/105454), 142.43 MiB | 10.25 MiB/s, done.
Resolving deltas: 100% (18567/18567), done.
From https://github.com/llvm/llvm-project
 * tag                   llvmorg-12.0.1-rc1 -> FETCH_HEAD
root@ryz2:~/asdf# git checkout FETCH_HEAD
Warning: you are leaving 1 commit behind, not connected to
any of your branches:

  d28af7c65 remove -fpch-codegen and -fpch-debuginfo from Clang 12.0 release notes

If you want to keep it by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> d28af7c65

HEAD is now at 328a6ec95 Force visibility of llvm::Any to external
root@ryz2:~/asdf# 
```

Reply:
> You can't use `git checkout $LLVM_VERSION`. Git fetch with depth=1 creates a short-lived ref.

It works fine if `$LLVM_VERSION` is a commit hash, but not for tags. But as fetching the tag along with the shallow clone isn't supported, then doing something with `FETCH_HEAD` probably is the way it has to be done anyway.

> There is another problem in your `prefer the code `. It will download every time all files from git (142 MB with compression). Here is example when I would like to change from 12.0.0 to 12.0.1-rc1

Sure, but I don't see how that's specific to my suggestion to the code?

Reply:
> Sure, but I don't see how that's specific to my suggestion to the code?

I was thinking something like that if something didn't work after updating llvm and you wanted to do a binary search it would take a while. 

> It works fine if `$LLVM_VERSION` is a commit hash, but not for tags

Why? It works for me. Tested also with git version `2.17.1` on docker `ubuntu:18.04`:
```console
root@ff06345fbb8c:~/asdf# history
    1  apt install git
    2  apt update
    3  apt install git
    4  cd
    5  mkdir asdf
    6  cd asdf
    7  git init 
    8  git remote add origin https://github.com/llvm/llvm-project.git
    9  git fetch --depth 1 origin "llvmorg-12.0.0"
   10  git -v
   11  git --version
   12  history
   13  git checkout FETCH_HEAD
   14  ls
   15  history
```

Reply:
> > Sure, but I don't see how that's specific to my suggestion to the code?
> 
> I was thinking something like that if something didn't work after updating llvm and you wanted to do a binary search it would take a while.
> 
> > It works fine if `$LLVM_VERSION` is a commit hash, but not for tags
> 
> Why? It works for me. Tested also with git version `2.17.1` on docker `ubuntu:18.04`:
> 
> ```
> root@ff06345fbb8c:~/asdf# history
>     1  apt install git
>     2  apt update
>     3  apt install git
>     4  cd
>     5  mkdir asdf
>     6  cd asdf
>     7  git init 
>     8  git remote add origin https://github.com/llvm/llvm-project.git
>     9  git fetch --depth 1 origin "llvmorg-12.0.0"
>    10  git -v
>    11  git --version
>    12  history
>    13  git checkout FETCH_HEAD
>    14  ls
>    15  history
> ```

I'm not saying `FETCH_HEAD` doesn't work, it does. I was saying that using `$LLVM_VERSION` instead of `FETCH_HEAD` for `git checkout` works after a shallow clone, if `$LLVM_VERSION` is a git commit hash, but not if it's a tag (since the shallow fetch doesn't get any tags associated with that commit).

Hence, as git behaves this way, I'm ok with the way of the PR looks right now.